### PR TITLE
HBASE-23756. [OpenTracing] Add OpenTracing dependency and helper methods.

### PR DESCRIPTION
**Pull request by jojochuang**
(1) Add OpenTracing dependencies, shade them appropriately.
(2) Add a few helper methods in TraceUtil, and added test for the helper methods.

### CODE REVIEW

**Comment by jojochuang** [comment]
This PR is not going to compile properly because of HBASE-23811. However I do want to kick this off early. Preferably this should be in a feature branch.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 29s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 2 new or modified test files.  |
||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 47s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   5m 14s |  master passed  |
| +1 :green_heart: |  compile  |   3m  0s |  master passed  |
| +1 :green_heart: |  checkstyle  |   2m 20s |  master passed  |
| +1 :green_heart: |  shadedjars  |   4m 37s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   4m 52s |  master passed  |
| +0 :ok: |  spotbugs  |   0m 25s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +0 :ok: |  findbugs  |   0m 40s |  branch/hbase-shaded no findbugs output file (findbugsXml.xml)  |
| +0 :ok: |  findbugs  |   0m 17s |  branch/hbase-shaded/hbase-shaded-client no findbugs output file (findbugsXml.xml)  |
| +0 :ok: |  findbugs  |   0m 14s |  branch/hbase-shaded/hbase-shaded-with-hadoop-check-invariants no findbugs output file (findbugsXml.xml)  |
| +0 :ok: |  findbugs  |   0m 16s |  branch/hbase-shaded/hbase-shaded-check-invariants no findbugs output file (findbugsXml.xml)  |
| +0 :ok: |  findbugs  |   0m 25s |  branch/hbase-shaded/hbase-shaded-testing-util no findbugs output file (findbugsXml.xml)  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 18s |  Maven dependency ordering for patch  |
| -1 :x: |  mvninstall  |   0m 27s |  root in the patch failed.  |
| -1 :x: |  compile  |   0m 12s |  root in the patch failed.  |
| -1 :x: |  javac  |   0m 12s |  root in the patch failed.  |
| -1 :x: |  checkstyle  |   0m  9s |  The patch fails to run checkstyle in root  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  xml  |   0m  9s |  The patch has no ill-formed XML file.  |
| -1 :x: |  shadedjars  |   0m 26s |  patch has 11 errors when building our shaded downstream artifacts.  |
| -1 :x: |  hadoopcheck  |   0m 12s |  The patch causes 10 errors with Hadoop v2.8.5.  |
| -1 :x: |  hadoopcheck  |   0m 24s |  The patch causes 10 errors with Hadoop v2.9.2.  |
| -1 :x: |  hadoopcheck  |   0m 36s |  The patch causes 10 errors with Hadoop v3.1.2.  |
| -1 :x: |  javadoc  |   0m 10s |  hbase-shaded in the patch failed.  |
| -1 :x: |  javadoc  |   0m 12s |  root in the patch failed.  |
| -1 :x: |  javadoc  |   0m  8s |  hbase-common in the patch failed.  |
| -1 :x: |  javadoc  |   0m 10s |  hbase-server in the patch failed.  |
| -1 :x: |  javadoc  |   0m  9s |  hbase-shaded-client in the patch failed.  |
| -1 :x: |  javadoc  |   0m  9s |  hbase-shaded-testing-util in the patch failed.  |
| -1 :x: |  findbugs  |   0m 12s |  hbase-shaded in the patch failed.  |
| -1 :x: |  findbugs  |   0m 13s |  root in the patch failed.  |
| -1 :x: |  findbugs  |   0m  8s |  hbase-common in the patch failed.  |
| -1 :x: |  findbugs  |   0m 10s |  hbase-server in the patch failed.  |
| +0 :ok: |  findbugs  |   0m 11s |  hbase-shaded/hbase-shaded-check-invariants has no data from findbugs  |
| -1 :x: |  findbugs  |   0m  8s |  hbase-shaded-client in the patch failed.  |
| -1 :x: |  findbugs  |   0m  9s |  hbase-shaded-testing-util in the patch failed.  |
| +0 :ok: |  findbugs  |   0m 12s |  hbase-shaded/hbase-shaded-with-hadoop-check-invariants has no data from findbugs  |
||| _ Other Tests _ |
| -1 :x: |  unit  |   0m 28s |  root in the patch failed.  |
| +1 :green_heart: |  asflicense  |   1m 20s |  The patch does not generate ASF License warnings.  |
|  |   |  51m  9s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.6 Server=19.03.6 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1202/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/1202 |
| Optional Tests | dupname asflicense checkstyle xml javac javadoc unit shadedjars hadoopcheck compile spotbugs findbugs hbaseanti |
| uname | Linux 5a11615a8340 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/Base-PreCommit-GitHub-PR_PR-1202/out/precommit/personality/provided.sh |
| git revision | master / 3e21dc73de |
| Default Java | 1.8.0_181 |
| mvninstall | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1202/1/artifact/out/patch-mvninstall-root.txt |
| compile | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1202/1/artifact/out/patch-compile-root.txt |
| javac | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1202/1/artifact/out/patch-compile-root.txt |
| checkstyle | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1202/1/artifact/out//home/jenkins/jenkins-slave/workspace/Base-PreCommit-GitHub-PR_PR-1202/out/maven-patch-checkstyle-root.txt |
| shadedjars | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1202/1/artifact/out/patch-shadedjars.txt |
| hadoopcheck | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1202/1/artifact/out/patch-javac-2.8.5.txt |
| hadoopcheck | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1202/1/artifact/out/patch-javac-2.9.2.txt |
| hadoopcheck | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1202/1/artifact/out/patch-javac-3.1.2.txt |
| javadoc | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1202/1/artifact/out/patch-javadoc-hbase-shaded.txt |
| javadoc | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1202/1/artifact/out/patch-javadoc-root.txt |
| javadoc | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1202/1/artifact/out/patch-javadoc-hbase-common.txt |
| javadoc | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1202/1/artifact/out/patch-javadoc-hbase-server.txt |
| javadoc | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1202/1/artifact/out/patch-javadoc-hbase-shaded_hbase-shaded-client.txt |
| javadoc | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1202/1/artifact/out/patch-javadoc-hbase-shaded_hbase-shaded-testing-util.txt |
| findbugs | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1202/1/artifact/out/patch-findbugs-hbase-shaded.txt |
| findbugs | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1202/1/artifact/out/patch-findbugs-root.txt |
| findbugs | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1202/1/artifact/out/patch-findbugs-hbase-common.txt |
| findbugs | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1202/1/artifact/out/patch-findbugs-hbase-server.txt |
| findbugs | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1202/1/artifact/out/patch-findbugs-hbase-shaded_hbase-shaded-client.txt |
| findbugs | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1202/1/artifact/out/patch-findbugs-hbase-shaded_hbase-shaded-testing-util.txt |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1202/1/artifact/out/patch-unit-root.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1202/1/testReport/ |
| Max. process+thread count | 96 (vs. ulimit of 10000) |
| modules | C: hbase-shaded . hbase-common hbase-server hbase-shaded/hbase-shaded-check-invariants hbase-shaded/hbase-shaded-client hbase-shaded/hbase-shaded-testing-util hbase-shaded/hbase-shaded-with-hadoop-check-invariants U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1202/1/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

File: hbase-common/src/main/java/org/apache/hadoop/hbase/trace/TraceUtil.java [hunk 1]
```diff
@@ -34,17 +51,24 @@
...
-      conf = new HBaseHTraceConfiguration(c);
-    }
+  public static void initTracer(Configuration c, String serviceName) {
+    if (!GlobalTracer.isRegistered()) {
+      io.jaegertracing.Configuration conf = io.jaegertracing.Configuration.fromEnv(serviceName);
+      io.opentracing.Tracer tracer = conf.getTracerBuilder().build();
 
-    if (tracer == null && conf != null) {
-      tracer = new Tracer.Builder("Tracer").conf(conf).build();
+      GlobalTracer.register(tracer);
```
Comment (ramkrish86) [inline hbase-common/src/main/java/org/apache/hadoop/hbase/trace/TraceUtil.java]: More of a question  - This happens at the init time - means the server start up time right? If the user wants to use jaeger tracing he can do it by setting some env variables. But if the same has to be unset what needs to be done? And also after the server starts up and then he needs to introduce jaeger tracing how can that be done?

Comment (jojochuang) [reply hbase-common/src/main/java/org/apache/hadoop/hbase/trace/TraceUtil.java]: I'd suggest you to browse the Jaeger's sampling doc: https://www.jaegertracing.io/docs/1.14/sampling/#collector-sampling-configuration
You can use env variable to control the sampling strategy. Or programmatically (at init). Finally, you can also define the sampling strategy at the central trace collector.

File: hbase-common/src/main/java/org/apache/hadoop/hbase/trace/TraceUtil.java [hunk 2]
```diff
@@ -34,17 +51,24 @@
...
   private TraceUtil() {
   }
 
-  public static void initTracer(Configuration c) {
-    if (c != null) {
-      conf = new HBaseHTraceConfiguration(c);
-    }
+  public static void initTracer(Configuration c, String serviceName) {
+    if (!GlobalTracer.isRegistered()) {
+      io.jaegertracing.Configuration conf = io.jaegertracing.Configuration.fromEnv(serviceName);
```
Comment (Apache9) [inline hbase-common/src/main/java/org/apache/hadoop/hbase/trace/TraceUtil.java:64]: Can we avoid referencing jaeger directly in code? I would expect that this just work like slf4j, when only need to depend on log4j when testing or at runtime, not compile time.

**Review by Apache9 — CHANGES_REQUESTED** [review]
Do you still work on this @jojochuang ? If not, I would like to have a try.

**Comment by Apache9** [comment]
We plan to go with open telemetry for now.

**Pull request closed without merging**